src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java

```diff
@@ -175,19 +178,52 @@ protected StarlarkBaseExternalContext(
     this.remoteExecutor = remoteExecutor;
     this.asyncTasks = new ArrayList<>();
     this.allowWatchingPathsOutsideWorkspace = allowWatchingPathsOutsideWorkspace;
+    this.executorService =
+        Executors.newThreadPerTaskExecutor(
+            Thread.ofVirtual()
+                .name("downloads-" + workingDirectory.getBaseName())
+                .factory());
+  }
+
+  /**
+   * Mark the evaluation using this context as otherwise successful. This is used to determine how
+   * to clean up resources in {@link #close()}.
+   */
+  public final void markSuccessful() {
+    wasSuccessful = true;
   }
 
-  public boolean ensureNoPendingAsyncTasks(EventHandler eventHandler, boolean forSuccessfulFetch) {
+  @Override
+  public final void close() throws EvalException, IOException {
+    // Cancel all pending async tasks.
+    boolean hadPendingItems = ensureNoPendingAsyncTasks();
+    // Wait for all (cancelled) async tasks to complete before cleaning up the working directory.
+    // This is necessary because downloads may still be in progress and could end up writing to the
+    // working directory during deletion, which would cause an error.
+    executorService.close();
+    if (shouldDeleteWorkingDirectory(wasSuccessful)) {
+      workingDirectory.deleteTree();
+    }
+    if (hadPendingItems && wasSuccessful) {
+      throw Starlark.errorf(
+          "Pending asynchronous work after %s finished execution",
+          getIdentifyingStringForLogging());
+    }
+  }
+
+  private boolean ensureNoPendingAsyncTasks() {
```

Comment:
maybe this should be called `cancelPendingAsyncTasks`, since it's actually not ensuring the tasks have stopped, anyway.